### ISSUE TITLE
[RFC] Replace MatMul with BatchMatMul

### ISSUE THREAD

**Issue by qcolombet** [pytorch/glow#1573]
TL;DR I suggest to replace MatMul with BatchMatMul, the semantic would be:
Inputs [b0,..., bx, m, k], [b0,...,bx, k, n]
Output: [b0, ..., bx, m, n]

We currently expand Batched matmul operators into a bunch of matmul operators with slicing, reshape and concats.

One of the main drawback from this representation is a bloat of the number of operators ending up in the IR and the underlying issue of compile time, memory consumption, hard to analysis sequences to name a few.

Instead, I would suggest to replace our matmul operator with a batched matmul operator. Indeed, matmuls are just a special case of a batch of 1.
Using this kind of representation would give each backend the freedom to unroll or use loops to realize the expected sequence.
Moreover, it would make the modeling of predication more natural, instead of having to do the same gymnastic with slicing and the kind.

**Comment by artemrakhov** [comment]
Could you please explain more on how it helps with predication?

In fr2en, we regularly do multiplication of activation (`batchSize x N` matrix) and weight (`N x M` matrix), resulting in `batchSize x M` matrix. So predication needs to happen on non-batch dimension (first). When we do Grad of such FC, then we'll generate MatMul of size `N x batchSize` * `batchSize x M`. This means that predication needs to happen on second dimension (`batchSize`).

**Comment by qcolombet** [comment]
> Could you please explain more on how it helps with predication?

TL;DR, I believe we want to use batchmatmul in your example with a broadcast semantic of RHS.

Sure!
Thanks for the example, that's not what I had in mind for batch matmul :).

For a true batch matmul, i.e., at least 3d tensors on both sides, the fact that we unroll the batch implies that we have to do the same thing with the predicates. I would expect this to result in very bad code gen.

In your example, predication as we envisioned it does not really fit. Indeed, looking at just the operator, matmul, we think we proceed only on batch, so the predicate would be only 1 boolean.
I'd argue that if we want to have proper predication in that case, we need to express this matmul at a batch matmul with broadcasting of the RHS.

We could imagine that the predicate applies to the inputs shapes, but this gets complicated very quickly.

**Comment by qcolombet** [comment]
> We could imagine that the predicate applies to the inputs shapes, but this gets complicated very quickly.

By that I meant the predicate has the same shape as the inputs.
The problem, then is that the operator itself needs to update the predicate content and shape for the next operator.

**Comment by artemrakhov** [comment]
What is broadcast semantic of RHS? Do you suggest to do it with `{batchSize, 1, N} * {1, N, M}` multiplication?


> In your example, predication as we envisioned it does not really fit.

Recurrent NN is the main use case and motivation for introducing predication. We have to design predicates in such a way, that fr2en would not waste any computation.
I thought that predicate would be a vector of `{batchSize}`, and for ops such as MatMul we'll also have `predicatePosition` parameter, to mark which dimension is predicated.

**Comment by qcolombet** [comment]
> What is broadcast semantic of RHS? 

For your example, `{batchSize, N} * {N * M}`, would become `{batchSize, 1, N} * {batchSize, N, M}`, RHS, is broadcasted over the missing "batched" dimension(s).
In terms of implementation this could be fairly efficient, because we actually don't have to broadcast the RHS, we only need to read the same address again and again.

> I thought that predicate would be a vector of {batchSize}, and for ops such as MatMul we'll also have predicatePosition parameter, to mark which dimension is predicated.

The thing that bothers me with such approach is that we have to track where the batch dimension is. For instance, imagine that we have a predicated transpose, then the resulting transpose will potential have to provide a new predicate for the result to be correctly predicated. On top of that if we start to compose predicates, we will have to do some gymnastic to make sure the predicates are on the same dimensions or we could end up with several predicated dimensions(?).

**Comment by qcolombet** [comment]
> When we do Grad of such FC, then we'll generate MatMul of size N x batchSize * batchSize x M. This means that predication needs to happen on second dimension (batchSize).

That actually illustrates my point of things being complicated pretty quickly. Indeed the predicate applies to the second dimension of the first input, but when we look at the formula, it seems strange it also doesn't apply to the first dimension of the second input. By keeping the batch size outside of the formula, things are easy to understand IMHO.

**Comment by qcolombet** [comment]
Hi @nadavrot, @artemrakhov Did we reach an agreement that we only need the BatchMatMul operator?

**Comment by artemrakhov** [comment]
I'm afraid we did not yet.

I feel that this is tightly coupled with 3 other questions:

1. Broadcast should not physically copy data.
1. Semantics of predication. It can happen not only on first dimension. As you correctly mentioned above, in backward (training) pass of seq2seq, predication happens on second dim of LHS which is also the first dim of RHS. As for multiple batch dims: I don't think this can happen.
1. Tiling MatMul. If you have `{batchSize, 1, N} * {batchSize, N, M}`, how do you run it efficiently? Doing `batchSize` of vector-to-matrix mults is not going to be efficient.

**Comment by qcolombet** [comment]
> I'm afraid we did not yet.

That's what I thought :).

Thanks @artemrakhov for summarizing the points to clarify.

> 1. Broadcast should not physically copy data.

Agreed, but that's an implementation detail. Meaning whatever we decide doesn't impact the high level modeling.

> 2. Semantics of predication. It can happen not only on first dimension. As you correctly mentioned above, in backward (training) pass of seq2seq, predication happens on second dim of LHS which is also the first dim of RHS. As for multiple batch dims: I don't think this can happen.

I have two answers to this:
A. We could drop predication for training
B. Can this happen if the batch size is on the side? Meaning if FC are lowered as BatchMatMul of `{batchSize, 1, N} * {batchSize, N, M}` wouldn't we get a gradient with the batchSize still on the first dimension?
E.g., something like batchMatMul `{batchSize, N, 1} * {batchSize, 1, M}`.

#A is not as silly as it may sound. For training we control the inputs, therefore it should be possible to massage them in such a way that predication won't actually get us anything.

> 3. Tiling MatMul. If you have {batchSize, 1, N} * {batchSize, N, M}, how do you run it efficiently? Doing batchSize of vector-to-matrix mults is not going to be efficient.

To me that's an implementation detail too. What I had in mind was that we could totally special case batchmatmul to one regular matmul at codegen time when we see batched vectors. Again this doesn't invalidate the high level modeling.

The thing I am trying to convey is we can be as clever as we want at codegen time to match/beat the performance of a single matmul, and that's orthogonal to the proposal. However, I believe BatchMatMul gives us a strictly more flexible semantic than MatMul, while making things easier like optimizing batch of MatMul.
